compiler/src/iree/compiler/Codegen/Dialect/MicroKernelOps.cpp

```diff
@@ -0,0 +1,324 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Dialect/MicroKernelOps.h"
+
+#include "iree/builtins/ukernel/exported_bits.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/FormatVariadic.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/OpImplementation.h"
+#include "mlir/Support/LLVM.h"
+
+// clang-format off
+#define GET_OP_CLASSES
+#include "iree/compiler/Codegen/Dialect/MicroKernelOps.cpp.inc" // IWYU pragma: keep
+// clang-format on
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Codegen {
+
+/// Returns true if the dimensions of ShapedType are compatible.
+static bool isShapedTypeDimCompatible(int64_t lhs, int64_t rhs) {
+  return lhs == ShapedType::kDynamic || rhs == ShapedType::kDynamic ||
+         lhs == rhs;
+}
+
+/// Returns true if the dimensions of ShapedType are compatible.
+static bool areShapesCompatible(ArrayRef<int64_t> lhs, ArrayRef<int64_t> rhs) {
+  if (lhs.size() != rhs.size()) {
+    return false;
+  }
+  return llvm::all_of(llvm::zip(lhs, rhs), [](std::tuple<int64_t, int64_t> it) {
+    return isShapedTypeDimCompatible(std::get<0>(it), std::get<1>(it));
+  });
+}
+
+/// Helper method to generate a function declaration at a module scope,
+/// and a call to that function
+FailureOr<func::CallOp> createFunctionCall(RewriterBase &rewriter,
+                                           Operation *op, StringRef fnName,
+                                           TypeRange callArgumentTypes,
+                                           TypeRange callReturnTypes,
+                                           ValueRange callOperands) {
+  FunctionType functionType =
+      rewriter.getFunctionType(callArgumentTypes, callReturnTypes);
+
+  // Create a declaration for the function type.
+  Location loc = op->getLoc();
+  auto moduleOp = SymbolTable::getNearestSymbolTable(op);
+  // Check for duplicates.
+  auto fnDecl = dyn_cast_or_null<func::FuncOp>(
+      SymbolTable::lookupSymbolIn(moduleOp, fnName));
+  if (fnDecl) {
+    if (fnDecl.getFunctionType() != functionType) {
+      return rewriter.notifyMatchFailure(
+          op,
+          llvm::formatv("mismatch in function type computed during lowering "
+                        "({0}) and already declared function ({1})",
+                        functionType, fnDecl.getFunctionType()));
+    }
+  } else {
+    OpBuilder::InsertionGuard g(rewriter);
+    rewriter.setInsertionPointToStart(&moduleOp->getRegion(0).front());
+    fnDecl = rewriter.create<func::FuncOp>(loc, fnName, functionType);
+    SymbolTable::setSymbolVisibility(fnDecl, SymbolTable::Visibility::Private);
+  }
```

Comment:
[optional] The below form could save us one level of indent.

```
if (!fnDecl) {
  ...
} else if (fn.Decl.getFunctionType() != functionType) {
  ...
}
```

Reply:
Thanks!